lib/auth/googleauth.ts

```diff
@@ -0,0 +1,594 @@
+/**
+ * Copyright 2014 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { exec } from 'child_process';
+import Compute from './computeclient';
+import IAMAuth from './iam';
+import JWTAccess from './jwtaccess';
+import JWTClient from './jwtclient';
+import JWT from './jwtclient';
+import OAuth2 from './oauth2client';
+import UserRefreshClient from './refreshclient';
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as util from 'util';
+import DefaultTransporter from '../transporters';
+
+export default class GoogleAuth {
+
+  public transporter: DefaultTransporter;
+
+  /**
+   * Caches a value indicating whether the auth layer is running on Google Compute Engine.
+   * @private
+   */
+  private _isGCE = false;
+
+  // Note:  this properly is only public to satisify unit tests.
+  // https://github.com/Microsoft/TypeScript/issues/5228
+  get isGCE() {
+    return this._isGCE;
+  }
+
+  /**
+   * Caches a value indicating whether we've checked whether the auth layer is running on
+   * Google Compute Engine.
+   * @private
+   */
+  private _checked_isGCE = false;
+
+  // Note:  this properly is only public to satisify unit tests.
+  // https://github.com/Microsoft/TypeScript/issues/5228
+  get checked_isGCE() {
+    return this._checked_isGCE;
+  }
+
+  public _cachedProjectId: string;
+  public cachedCredential = null;
+
+  protected JWTClient = JWTClient;
+  protected ComputeClient = Compute;
+
+  /**
+   * Convenience field mapping in the IAM credential type.
+   */
+  public IAMAuth = IAMAuth;
+
+  /**
+   * Convenience field mapping in the Compute credential type.
+   */
+  public Compute = Compute;
+
+  /**
+   * Convenience field mapping in the JWT credential type.
+   */
+  public JWT = JWT;
```

Comment:
I am trying to avoid changes to the public API surface if at all possible.  I'm completely open to changing this later, but I wanted to focus on the same API contract and just converting to TypeScript.  Let me know if you disagree.